### ISSUE TITLE
Workspace not initialized

### ISSUE THREAD

**Issue by smiller171** [hashicorp/vscode-terraform#329]
## Workaround:
For anyone who comes across this issue, as long as your TF is in the root of your project, you're likely seeing this issue within a child module that doesn't declare it's own providers, instead relying on inheritance from a parent module. If this is the case, a workaround that works for me is to define the `required_providers` so that Terraform knows what providers to cache when you `init`.

example:
```hcl
terraform {
  required_providers {
    aws = "~> 2.61.0"
  }
}
```

## Original issue description:
When trying to enable the language server with the new `2.0.0-rc1` version, I get a "Workspace not initialized" error, which is not resolved by running `terraform init`

```
Workspace not initialized. Please run `terraform init` in /Users/smiller/code/terraform-anycloud-template
```
```
Launching language server: /Users/smiller/.vscode/extensions/hashicorp.terraform-2.0.0-rc.1/lsp/terraform-ls serve
2020/06/04 15:37:53 service.go:63: Preparing new session ...
2020/06/04 15:37:53 langserver.go:79: Starting server (pid 53425) ...
2020/06/04 15:37:53 server.go:469: Received 1 new requests
2020/06/04 15:37:53 server.go:165: Processing 1 requests
2020/06/04 15:37:53 server.go:223: Checking request for "initialize": {"processId":52705,"rootPath":"/Users/smiller/code/terraform-anycloud-template","rootUri":"file:///Users/smiller/code/terraform-anycloud-template","capabilities":{"workspace":{"applyEdit":true,"workspaceEdit":{"documentChanges":true,"resourceOperations":["create","rename","delete"],"failureHandling":"textOnlyTransactional"},"didChangeConfiguration":{"dynamicRegistration":true},"didChangeWatchedFiles":{"dynamicRegistration":true},"symbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},"executeCommand":{"dynamicRegistration":true},"configuration":true,"workspaceFolders":true},"textDocument":{"publishDiagnostics":{"relatedInformation":true},"synchronization":{"dynamicRegistration":true,"willSave":true,"willSaveWaitUntil":true,"didSave":true},"completion":{"dynamicRegistration":true,"contextSupport":true,"completionItem":{"snippetSupport":true,"commitCharactersSupport":true,"documentationFormat":["markdown","plaintext"],"deprecatedSupport":true,"preselectSupport":true},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"hover":{"dynamicRegistration":true,"contentFormat":["markdown","plaintext"]},"signatureHelp":{"dynamicRegistration":true,"signatureInformation":{"documentationFormat":["markdown","plaintext"],"parameterInformation":{"labelOffsetSupport":true}}},"definition":{"dynamicRegistration":true,"linkSupport":true},"references":{"dynamicRegistration":true},"documentHighlight":{"dynamicRegistration":true},"documentSymbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"hierarchicalDocumentSymbolSupport":true},"codeAction":{"dynamicRegistration":true,"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"codeLens":{"dynamicRegistration":true},"formatting":{"dynamicRegistration":true},"rangeFormatting":{"dynamicRegistration":true},"onTypeFormatting":{"dynamicRegistration":true},"rename":{"dynamicRegistration":true,"prepareSupport":true},"documentLink":{"dynamicRegistration":true},"typeDefinition":{"dynamicRegistration":true,"linkSupport":true},"implementation":{"dynamicRegistration":true,"linkSupport":true},"colorProvider":{"dynamicRegistration":true},"foldingRange":{"dynamicRegistration":true,"rangeLimit":5000,"lineFoldingOnly":true},"declaration":{"dynamicRegistration":true,"linkSupport":true}}},"trace":"off","workspaceFolders":[{"uri":"file:///Users/smiller/code/terraform-anycloud-template","name":"terraform-anycloud-template"}]}
2020/06/04 15:37:53 rpc_logger.go:29: Incoming request for "initialize" (ID 0): {"processId":52705,"rootPath":"/Users/smiller/code/terraform-anycloud-template","rootUri":"file:///Users/smiller/code/terraform-anycloud-template","capabilities":{"workspace":{"applyEdit":true,"workspaceEdit":{"documentChanges":true,"resourceOperations":["create","rename","delete"],"failureHandling":"textOnlyTransactional"},"didChangeConfiguration":{"dynamicRegistration":true},"didChangeWatchedFiles":{"dynamicRegistration":true},"symbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},"executeCommand":{"dynamicRegistration":true},"configuration":true,"workspaceFolders":true},"textDocument":{"publishDiagnostics":{"relatedInformation":true},"synchronization":{"dynamicRegistration":true,"willSave":true,"willSaveWaitUntil":true,"didSave":true},"completion":{"dynamicRegistration":true,"contextSupport":true,"completionItem":{"snippetSupport":true,"commitCharactersSupport":true,"documentationFormat":["markdown","plaintext"],"deprecatedSupport":true,"preselectSupport":true},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"hover":{"dynamicRegistration":true,"contentFormat":["markdown","plaintext"]},"signatureHelp":{"dynamicRegistration":true,"signatureInformation":{"documentationFormat":["markdown","plaintext"],"parameterInformation":{"labelOffsetSupport":true}}},"definition":{"dynamicRegistration":true,"linkSupport":true},"references":{"dynamicRegistration":true},"documentHighlight":{"dynamicRegistration":true},"documentSymbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"hierarchicalDocumentSymbolSupport":true},"codeAction":{"dynamicRegistration":true,"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"codeLens":{"dynamicRegistration":true},"formatting":{"dynamicRegistration":true},"rangeFormatting":{"dynamicRegistration":true},"onTypeFormatting":{"dynamicRegistration":true},"rename":{"dynamicRegistration":true,"prepareSupport":true},"documentLink":{"dynamicRegistration":true},"typeDefinition":{"dynamicRegistration":true,"linkSupport":true},"implementation":{"dynamicRegistration":true,"linkSupport":true},"colorProvider":{"dynamicRegistration":true},"foldingRange":{"dynamicRegistration":true,"rangeLimit":5000,"lineFoldingOnly":true},"declaration":{"dynamicRegistration":true,"linkSupport":true}}},"trace":"off","workspaceFolders":[{"uri":"file:///Users/smiller/code/terraform-anycloud-template","name":"terraform-anycloud-template"}]}
2020/06/04 15:37:53 exec.go:182: Starting /usr/local/bin/terraform ["terraform" "version"] in "/var/folders/43/thr3y6c55s718zjvk__hcns4hg_30g/T/"...
2020/06/04 15:37:53 exec.go:148: Waiting for command to finish ...
2020/06/04 15:37:54 exec.go:174: terraform run (/usr/local/bin/terraform ["terraform" "version"], in "/var/folders/43/thr3y6c55s718zjvk__hcns4hg_30g/T/", pid 53426) finished with exit code 0
2020/06/04 15:37:54 initialize.go:90: Found compatible Terraform version (0.12.26) at /usr/local/bin/terraform
2020/06/04 15:37:54 schema_storage.go:334: Adding workspace for watching: "/Users/smiller/code/terraform-anycloud-template"
2020/06/04 15:37:54 watcher.go:88: Adding "/Users/smiller/code/terraform-anycloud-template/.terraform/plugins/darwin_amd64/lock.json" for watching...
2020/06/04 15:37:54 schema_storage.go:124: Acquiring semaphore before retrieving schema for "/Users/smiller/code/terraform-anycloud-template" ...
2020/06/04 15:37:54 schema_storage.go:133: Retrieving schemas for "/Users/smiller/code/terraform-anycloud-template" ...
2020/06/04 15:37:54 exec.go:182: Starting /usr/local/bin/terraform ["terraform" "providers" "schema" "-json"] in "/Users/smiller/code/terraform-anycloud-template"...
2020/06/04 15:37:54 rpc_logger.go:45: Error for "initialize" (ID 0): [-32098] Workspace not initialized. Please run `terraform init` in /Users/smiller/code/terraform-anycloud-template
2020/06/04 15:37:54 server.go:204: Completed 1 requests [1.091489042s elapsed]
[Error - 3:37:54 PM] Server initialization failed.
  Message: Workspace not initialized. Please run `terraform init` in /Users/smiller/code/terraform-anycloud-template
  Code: -32098 
2020/06/04 15:37:54 server.go:469: Received 1 new requests
2020/06/04 15:37:54 exec.go:148: Waiting for command to finish ...
2020/06/04 15:37:54 server.go:165: Processing 1 requests
2020/06/04 15:37:54 server.go:223: Checking request for "shutdown": 
2020/06/04 15:37:54 rpc_logger.go:29: Incoming request for "shutdown" (ID 1): 
2020/06/04 15:37:54 rpc_logger.go:50: Response to "shutdown" (ID 1): null
2020/06/04 15:37:54 server.go:204: Completed 1 requests [132.937µs elapsed]
2020/06/04 15:37:54 server.go:469: Received 1 new requests
2020/06/04 15:37:54 server.go:165: Processing 1 requests
2020/06/04 15:37:54 server.go:223: Checking request for "exit": 
2020/06/04 15:37:54 rpc_logger.go:29: Incoming notification for "exit": 
2020/06/04 15:37:54 schema_storage.go:116: error obtaining schemas: Unable to retrieve schemas: failed to get schemas: Execution of ["terraform" "providers" "schema" "-json"] canceled.
signal: killed
```

**Comment by paultyng** [comment]
Yes, we plan to add an initialization workflow where the extension prompts. We can use this as the tracking issue (I'm still working my way through issue cleanup and creation).

**Comment by forstermatth** [comment]
I'm also getting a similar error for a multi-repository workspace. One repo is not a terraform project, and the other repo has multiple terraform root folders within it.

```
Launching language server: /Users/matt/.vscode/extensions/hashicorp.terraform-2.0.0-rc.1/lsp/terraform-ls serve
2020/06/04 14:28:50 service.go:63: Preparing new session ...
2020/06/04 14:28:50 langserver.go:79: Starting server (pid 81445) ...
2020/06/04 14:28:50 server.go:469: Received 1 new requests
2020/06/04 14:28:50 server.go:165: Processing 1 requests
2020/06/04 14:28:50 server.go:223: Checking request for "initialize": {"processId":81211,"rootPath":"/Users/matt/non-terraform-project","rootUri":"file:///Users/matt/non-terraform-project","capabilities":{"workspace":{"applyEdit":true,"workspaceEdit":{"documentChanges":true,"resourceOperations":["create","rename","delete"],"failureHandling":"textOnlyTransactional"},"didChangeConfiguration":{"dynamicRegistration":true},"didChangeWatchedFiles":{"dynamicRegistration":true},"symbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},"executeCommand":{"dynamicRegistration":true},"configuration":true,"workspaceFolders":true},"textDocument":{"publishDiagnostics":{"relatedInformation":true},"synchronization":{"dynamicRegistration":true,"willSave":true,"willSaveWaitUntil":true,"didSave":true},"completion":{"dynamicRegistration":true,"contextSupport":true,"completionItem":{"snippetSupport":true,"commitCharactersSupport":true,"documentationFormat":["markdown","plaintext"],"deprecatedSupport":true,"preselectSupport":true},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"hover":{"dynamicRegistration":true,"contentFormat":["markdown","plaintext"]},"signatureHelp":{"dynamicRegistration":true,"signatureInformation":{"documentationFormat":["markdown","plaintext"],"parameterInformation":{"labelOffsetSupport":true}}},"definition":{"dynamicRegistration":true,"linkSupport":true},"references":{"dynamicRegistration":true},"documentHighlight":{"dynamicRegistration":true},"documentSymbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"hierarchicalDocumentSymbolSupport":true},"codeAction":{"dynamicRegistration":true,"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"codeLens":{"dynamicRegistration":true},"formatting":{"dynamicRegistration":true},"rangeFormatting":{"dynamicRegistration":true},"onTypeFormatting":{"dynamicRegistration":true},"rename":{"dynamicRegistration":true,"prepareSupport":true},"documentLink":{"dynamicRegistration":true},"typeDefinition":{"dynamicRegistration":true,"linkSupport":true},"implementation":{"dynamicRegistration":true,"linkSupport":true},"colorProvider":{"dynamicRegistration":true},"foldingRange":{"dynamicRegistration":true,"rangeLimit":5000,"lineFoldingOnly":true},"declaration":{"dynamicRegistration":true,"linkSupport":true}}},"trace":"off","workspaceFolders":[{"uri":"file:///Users/matt/non-terraform-project","name":"images"},{"uri":"file:///Users/matt/terraform-project","name":"operations"}]}
2020/06/04 14:28:50 rpc_logger.go:29: Incoming request for "initialize" (ID 0): {"processId":81211,"rootPath":"/Users/matt/non-terraform-project","rootUri":"file:///Users/matt/non-terraform-project","capabilities":{"workspace":{"applyEdit":true,"workspaceEdit":{"documentChanges":true,"resourceOperations":["create","rename","delete"],"failureHandling":"textOnlyTransactional"},"didChangeConfiguration":{"dynamicRegistration":true},"didChangeWatchedFiles":{"dynamicRegistration":true},"symbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},"executeCommand":{"dynamicRegistration":true},"configuration":true,"workspaceFolders":true},"textDocument":{"publishDiagnostics":{"relatedInformation":true},"synchronization":{"dynamicRegistration":true,"willSave":true,"willSaveWaitUntil":true,"didSave":true},"completion":{"dynamicRegistration":true,"contextSupport":true,"completionItem":{"snippetSupport":true,"commitCharactersSupport":true,"documentationFormat":["markdown","plaintext"],"deprecatedSupport":true,"preselectSupport":true},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"hover":{"dynamicRegistration":true,"contentFormat":["markdown","plaintext"]},"signatureHelp":{"dynamicRegistration":true,"signatureInformation":{"documentationFormat":["markdown","plaintext"],"parameterInformation":{"labelOffsetSupport":true}}},"definition":{"dynamicRegistration":true,"linkSupport":true},"references":{"dynamicRegistration":true},"documentHighlight":{"dynamicRegistration":true},"documentSymbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"hierarchicalDocumentSymbolSupport":true},"codeAction":{"dynamicRegistration":true,"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"codeLens":{"dynamicRegistration":true},"formatting":{"dynamicRegistration":true},"rangeFormatting":{"dynamicRegistration":true},"onTypeFormatting":{"dynamicRegistration":true},"rename":{"dynamicRegistration":true,"prepareSupport":true},"documentLink":{"dynamicRegistration":true},"typeDefinition":{"dynamicRegistration":true,"linkSupport":true},"implementation":{"dynamicRegistration":true,"linkSupport":true},"colorProvider":{"dynamicRegistration":true},"foldingRange":{"dynamicRegistration":true,"rangeLimit":5000,"lineFoldingOnly":true},"declaration":{"dynamicRegistration":true,"linkSupport":true}}},"trace":"off","workspaceFolders":[{"uri":"file:///Users/matt/non-terraform-project","name":"images"},{"uri":"file:///Users/matt/terraform-project","name":"operations"}]}
2020/06/04 14:28:50 exec.go:182: Starting /Users/matt/.asdf/shims/terraform ["terraform" "version"] in "/var/folders/_m/2vc1tn9j7ng2d5kwkqs73zz80000gn/T/"...
2020/06/04 14:28:50 exec.go:148: Waiting for command to finish ...
2020/06/04 14:28:51 exec.go:174: terraform run (/Users/matt/.asdf/shims/terraform ["terraform" "version"], in "/var/folders/_m/2vc1tn9j7ng2d5kwkqs73zz80000gn/T/", pid 81446) finished with exit code 0
2020/06/04 14:28:51 initialize.go:90: Found compatible Terraform version (0.12.21) at /Users/matt/.asdf/shims/terraform
2020/06/04 14:28:51 schema_storage.go:334: Adding workspace for watching: "/Users/matt/non-terraform-project"
2020/06/04 14:28:51 watcher.go:88: Adding "/Users/matt/non-terraform-project/.terraform/plugins/darwin_amd64/lock.json" for watching...
2020/06/04 14:28:51 schema_storage.go:124: Acquiring semaphore before retrieving schema for "/Users/matt/non-terraform-project" ...
2020/06/04 14:28:51 schema_storage.go:133: Retrieving schemas for "/Users/matt/non-terraform-project" ...
2020/06/04 14:28:51 exec.go:182: Starting /Users/matt/.asdf/shims/terraform ["terraform" "providers" "schema" "-json"] in "/Users/matt/non-terraform-project"...
2020/06/04 14:28:51 rpc_logger.go:45: Error for "initialize" (ID 0): [-32098] Workspace not initialized. Please run `terraform init` in /Users/matt/non-terraform-project
2020/06/04 14:28:51 server.go:204: Completed 1 requests [234.295811ms elapsed]
[Error - 2:28:51 PM] Server initialization failed.
  Message: Workspace not initialized. Please run `terraform init` in /Users/matt/non-terraform-project
  Code: -32098 
2020/06/04 14:28:51 exec.go:148: Waiting for command to finish ...
2020/06/04 14:28:51 server.go:469: Received 1 new requests
2020/06/04 14:28:51 server.go:165: Processing 1 requests
2020/06/04 14:28:51 server.go:223: Checking request for "shutdown": 
2020/06/04 14:28:51 rpc_logger.go:29: Incoming request for "shutdown" (ID 1): 
2020/06/04 14:28:51 rpc_logger.go:50: Response to "shutdown" (ID 1): null
2020/06/04 14:28:51 server.go:204: Completed 1 requests [159.085µs elapsed]
2020/06/04 14:28:51 server.go:469: Received 1 new requests
2020/06/04 14:28:51 server.go:165: Processing 1 requests
2020/06/04 14:28:51 server.go:223: Checking request for "exit": 
2020/06/04 14:28:51 rpc_logger.go:29: Incoming notification for "exit": 
2020/06/04 14:28:51 schema_storage.go:116: error obtaining schemas: Unable to retrieve schemas: failed to get schemas: Execution of ["terraform" "providers" "schema" "-json"] canceled.
signal: killed
```

**Comment by smiller171** [comment]
> Yes, we plan to add an initialization workflow where the extension prompts. We can use this as the tracking issue (I'm still working my way through issue cleanup and creation).

This isn't just a matter of there not being a built-in workflow. It's totally impossible to resolve this. It instructs me to run `terraform init` but it's definitely initialized and not working I've tried running `init` several times to no avail.

**Comment by paultyng** [comment]
Yeah, unfortunately multi-folder is not yet supported by the LS. There is a possibility of doing multiple LS's per folder, or supporting multiple folders in a single LS process, and I don't know that we have chosen a direction yet, you can track that in #330

**Comment by paultyng** [comment]
@smiller171 what Terraform version are you on? (we should really output that in the logs probably cc @radeksimko )

**Comment by smiller171** [comment]
well, guess it'll be a while before the new language server is even remotely ready for me then.

unfortunately I can't even get the workflow working for configuring a different language server so I can keep using the old one.

@paultyng The run in the logs here was `0.12.26`. I use tfenv and have tried several different versions at this point.

**Comment by muram** [comment]
Not sure why, but I only see the old version of the extension. Uninstalled -> reinstalled the one from HashiCorp
![image](https://user-images.githubusercontent.com/20080955/83808637-c3d5a180-a67a-11ea-933f-937ec1abc3db.png)

**Comment by smiller171** [comment]
@muram that'd be because this new version isn't published yet, you need to sideload it.

Also your comment is totally irrelevant to this issue.

**Comment by paultyng** [comment]
@muram you'll need to manually download and install it from the GitHub [releases page](https://github.com/hashicorp/vscode-terraform/releases). It won't hit the marketplace until next week or the week after most likely.

**Comment by muram** [comment]
Got it, thank you @smiller171 @paultyng and sorry for hijacking this thread

**Comment by forstermatth** [comment]
@paultyng 

> @smiller171 what Terraform version are you on? (we should really output that in the logs probably cc @radeksimko )

The version is output in the logs, From @smiller171 output;
> 2020/06/04 15:37:54 initialize.go:90: Found compatible Terraform version (0.12.26) at /usr/local/bin/terraform

**Comment by paultyng** [comment]
@smiller171 I really appreciate you trying this and providing feedback and understand if its possibly too rough/early for you. 

One more data collection thing, can you manually run the command referenced in that output log: `terraform providers schema -json`.

My guess is this is `tfenv` related, I don't have that setup locally, but I'll spin up an environment to test it shortly.

**Comment by paultyng** [comment]
You can manually override how `terraform` is invoked in the LS in settings by modifying the `args` array. It should start with the `serve` command but then can use the following flags:

```shell
Usage: terraform-ls serve [options]

Starts the Language Server

Options:

  -log-file string
        path to a file to log into with support for variables (e.g. Timestamp, Pid, Ppid) via Go template syntax {{.VarName}}
  -port int
        port number to listen on (turns server into TCP mode)
  -tf-exec string
        path to Terraform binary
  -tf-exec-timeout string
        Overrides Terraform execution timeout (e.g. 30s)
  -tf-log-file string
        path to a file for Terraform executions to be logged into with support for variables (e.g. Timestamp, Pid, Ppid) via Go template syntax {{.VarName}}
```

**Comment by smiller171** [comment]
@paultyng 

```
default ❯ terraform providers schema -json
{"format_version":"0.1"}
```

This is in a pretty small module.

**Comment by radeksimko** [comment]
Just to provide some explanation/background, this error

> Workspace not initialized. Please run `terraform init`

would be returned when this path doesn't exist:

```go
filepath.Join(dir, ".terraform",
	"plugins",
	runtime.GOOS+"_"+runtime.GOARCH,
	"lock.json")
```
where `dir` is the current working directory, e.g. `./.terraform/plugins/darwin_amd64/lock.json` on macOS.

This is the default path to the plugin cache lock file; an implementation detail of how Terraform keeps track of providers locally after they are downloaded via `terraform init`.

This is admittedly a very simplified way of detecting whether a workspace was initialized and as such is has some known flaws:
 - The plugin cache may not be stored in the CWD, this may be because:
   - you use Terragrunt where plugin caches are decoupled from their directories - see https://github.com/hashicorp/terraform-ls/issues/32#issuecomment-628894419
   - you pass `TF_PLUGIN_CACHE_DIR` to your usual Terraform commands in some other ways - see https://github.com/hashicorp/terraform-ls/issues/24
   - you are simply editing files within a nested module folder, which would normally never be `init`d in its own directory, but rather from the one where the configuration references it - https://github.com/hashicorp/terraform-ls/issues/32 (Local Modules section) describes that situation
 - The plugin cache location/structure differs in 0.13 - which is tracked under https://github.com/hashicorp/terraform-ls/issues/111

All of the above cases are valid use cases we want to support, just bear in mind that this is an initial release, so we did take some shortcuts (including _not_ supporting the above yet) to gather early feedback and better gauge the priorities based on it.

With that in mind - **does any of the above describe your situation**? If not - can you please check if the lock file exists in the root of the directory you are opening in VSCode under `.terraform`? Also do you mind sharing any more details about where and how you `init` your folder? e.g. do you use any Terraform wrapper? Does it pass any particular flags or environment variables down to Terraform?

Thank you all for sharing this early feedback!

**Comment by marksmall** [comment]
I made some comments in #275 and was asked to look here for one of the problems I came up against. I think this explains why `code` thought the workspace wasn't initialised. The project I used is not a `terraform` project as such, but an application, which uses `terraform` to manage the **deployment**. So, our terraform code is not at the top-level, rather 
```
PROJ_ROOT/
    terraform/
        deploy/
            *.tf
```

This explains the problem I was seeing.

**Comment by radeksimko** [comment]
@marksmall Understood, that is a valid use case we intend to support. Have you tried opening the `deploy` folder directly as a project? I'm not saying that should be a long-term solution, but a workaround?

Relatedly am I right to say that you _would_ normally run `terraform init` inside the `deploy` folder?

**Comment by marksmall** [comment]
@radeksimko That is probably what I'll end up doing as a workaround while this problem is being fixed. It's not a big deal for me in the grand scheme of things.

As to you your question on `terraform init`, you are correct, I would run that in the `deploy` folder.

**Comment by smiller171** [comment]
@radeksimko I don't think any of these describe my situation. Is it possible that lock file is not created if no plugins are in use?

**Comment by radeksimko** [comment]
@smiller171 
> Is it possible that lock file is not created if no plugins are in use?

Yes, that is certainly possible. `terraform init` won't create any plugin cache if there aren't any plugins defined in the config.

I suppose that use case would be described as _"I'm starting from scratch and don't have any code yet"_? Which is also a totally valid (and probably common) use case we want to support. One way we can support that is by not erroring when schema isn't available on initialization as per https://github.com/hashicorp/terraform-ls/issues/83

**Comment by smiller171** [comment]
> I suppose that use case would be described as _"I'm starting from scratch and don't have any code yet"_? Which is also a totally valid (and common) use case we want to support. One way we can support that is by not erroring when schema isn't available on initialization as per [hashicorp/terraform-ls#83](https://github.com/hashicorp/terraform-ls/issues/83)

I think there's only one plugin we ever use, and it's an in-house plugin that we'll be deprecating once custom validation is in a good place. We have more than 70 terraform repos and I think that's the only plugin we use in any of them.

**Comment by radeksimko** [comment]
@smiller171 If there is one plugin then there should be a plugin cache and therefore a lock file.

Can you share some details about how/where you run `terraform init` then? Do you use any wrapper like Terragrunt, or do you pass any particular flags or ENV variables to `terraform init`? Can you check `ls -la` output after `terraform init`? Is there `.terraform` folder at all?

> I think there's only one plugin we ever use, and it's an in-house plugin that we'll be deprecating once custom validation is in a good place.

I'm having trouble understanding how could you use Terraform without _any_ plugin - it doesn't seem very useful on its own? So just to clarify - when I say plugin I mean a provider (like any provider from https://github.com/terraform-providers/). Are we on the same page? 😅

**Comment by smiller171** [comment]
@radeksimko ok, I didn't realize that providers show up as plugins. This is still an issue for nearly all of our modules because it only creates the `.terraform/plugins` dir if you explicitly call the provider, not if you're expecting the provider to be implicitly passed by the parent.

When I explicitly call the provider and init the directory and lockfile are created, but this isn't ideal within a child module.

> In all cases it is recommended to keep explicit provider configurations only in the root module and pass them (whether implicitly or explicitly) down to descendent modules. This avoids the provider configurations from being "lost" when descendent modules are removed from the configuration. It also allows the user of a configuration to determine which providers require credentials by inspecting only the root module.

**Comment by smiller171** [comment]
ok, I've just realized that the plugins lock is also created if I add the providers I'm using to `required_providers` (which is best practice and I just hadn't here) This at least unblocks me but since Terraform doesn't enforce this it creates problems for the functionality of the language server.

**Comment by smiller171** [comment]
after defining the provider in `required_providers` I tried to enable the language server again, and it asked me to update. I'm now getting:

```
Launching language server: /Users/smiller/.vscode/extensions/hashicorp.terraform-2.0.0-rc.1/lsp/terraform-ls serve
[Error - 11:28:10 AM] Connection to server is erroring. Shutting down server.
[Error - 11:28:10 AM] Connection to server is erroring. Shutting down server.
[Error - 11:28:10 AM] Connection to server is erroring. Shutting down server.
[Error - 11:28:10 AM] Connection to server is erroring. Shutting down server.
```

**Comment by smiller171** [comment]
I've manually rolled back to `0.3.0` instead of `0.3.1` of the language server and now it seems to be working.

I tried downloading `0.3.1` directly from the releases page and it won't even unzip

**Comment by radeksimko** [comment]
Thank you for describing all these steps along the way @smiller171 !

Supporting nested modules is indeed something we want to support - that falls under hashicorp/terraform-ls#32 (Local Modules section) as mentioned earlier and we're trying to prioritize solution for that use case.

As for the `[Error - 11:28:10 AM] Connection to server is erroring. Shutting down server.` error: Are you by any chance using macOS Catalina?

I ran into the exact same issue earlier myself and managed to basically fix it by `rm` ing that upgraded binary. I'm unsure what the root cause is yet, but I remember seeing differences in extended attributes (as per `xattr` output) between the "upgraded" binary and one I would download and unzip myself via browser.

> I tried downloading 0.3.1 directly from the releases page and it won't even unzip

We haven't changed anything about the release process or zipping between 0.3.0 and 0.3.1 and I did manage to fix the above problem by just `rm`ing that upgraded binary as displayed by VSCode - in your case that would be

```
rm -f /Users/smiller/.vscode/extensions/hashicorp.terraform-2.0.0-rc.1/lsp/terraform-ls
```
and then try restarting VSCode - the extension should download and install a fresh `0.3.1`.

**Comment by smiller171** [comment]
@radeksimko Thanks, removing `terraform-ls` and letting the extension re-install it worked (And I am on Catalina) It's the zip in the GitHub release that I tried to manually download that I can't unzip. The zip automatically downloaded by the extension is fine.

**Comment by radeksimko** [comment]
If you mean [this particular archive](https://github.com/hashicorp/terraform-ls/releases/download/v0.3.1/terraform-ls_0.3.1_darwin_amd64.zip) then that also works for me.

Just for sanity, did you check the archive was downloaded correctly?

```
$ shasum -a256 terraform-ls_0.3.1_darwin_amd64.zip
```
```
b972aaf1c4f40fc1aabf95266171eed97efd62e5f947221601d201d59187aa54
```

We also provide instructions on our Security page on how to verify the archive signature, in case you want to do that: https://www.hashicorp.com/security (`Archive Checksum Verification` section)

**Comment by smiller171** [comment]
I've updated the top comment to communicate the workaround of using `required_providers` for anyone who finds this issue, but I very much think this should be much smoother before public release.

**Comment by paultyng** [comment]
My working theory here (unfortunately not on Catalina) is that overwriting a binary functions somehow differently than deleting it and replacing it, but I'm trying to see if there is any literature on this. We can probably just update the extension to explicitly delete first though regardless as that seems to make this function correctly.

**Comment by J0F3** [comment]
>Have you tried opening the `deploy` folder directly as a project? I'm not saying that should be a long-term solution, but a workaround?
> 
@radeksimko Sorry, but that is not really a workaround in my case. This would make other configuration we have in the project like for VS Code Tasks oder Dev Container useless. Furthermore the Terraform code makes actually only sense under the context of the wohle project. Opening only the Terraform folder as a project removes this context and would make editing the terraform code just more complicated and harder that it must be.

I honestly can not understand why you have changed the behavior of the extension so fundamentally while I have not even a chance to stay at the old version (which you have took obviously over now) and nothing is documented in the release notes.
The current version is for me absolutely useless now. Please see also my comment on #366

**Comment by smiller171** [comment]
@J0F3 There is some documentation in the release notes on how to point to an alternate language server, so you can manually download the old one and use it.

This does feel extremely rushed though. I totally get releasing an RC for feedback but it's absurd to me that they would drop the actual 2.0.0 release in this state.

**Comment by J0F3** [comment]
@smiller171 Ah thanks for the hint. Didn't saw that.  Then I'll may give that a shot.  

And yeah I fully agree that when this would be a Preview version then it would be totally fine. But as it updates the extension automatically to this version and drops the old one completely this is absolutely no fun. 
Even the "pining" to the old version as described in the Release notes does currently not work.

**Comment by nbetm** [comment]
I found a simple workaround that works well for me. I added a single `terraform-ls.tf` file at the root of the project with only a list of providers I use.

**Comment by smiller171** [comment]
> Even the "pining" to the old version as described in the Release notes does currently not work.

If that's the case it's a vscode bug and not at all something Hashicorp should have to think overmuch about

**Comment by dimisjim** [comment]
Hashicorp took the initiative to work on this project. Of course they need to think this through.

**Comment by J0F3** [comment]
@smiller171 
Hmm, I don't think so. It works for other extensions but not for this one because simply no older version are found/published in the marketplace. 
#372 describes it. At least we can now manually download the old version...

**Comment by smiller171** [comment]
oh...that's not good...

**Comment by benmontour-wf** [comment]
I have things structured in the Terragrunt style, with an infrastructure-live repo that contains only terragrunt.hcl files that reference modules, and a modules repo that contains said Terraform modules. I used to use this extension to keep those modules and hcl files properly formatted and linted. I don't ever run Terraform commands from my local Terraform folders. I have automation that pulls down repos and runs commands as part of it's build/deploy. This extension has basically become useless for me now, since I don't have the root modules folder init'd.

**Comment by mgreen-sm** [comment]
@radeksimko: 

>   * you pass `TF_PLUGIN_CACHE_DIR` to your usual Terraform commands in some other ways - see [hashicorp/terraform-ls#24](https://github.com/hashicorp/terraform-ls/issues/24)

This would be our use case. We use our own variant of [tfutils/tfscaffold](https://github.com/tfutils/tfscaffold), which does indeed set the TF_PLUGIN_CACHE_DIR to something other than ./terraform by default (due to it cd-ing into a component directory).

So it would be good to see the option to provide the option to specify the directory, rather than just auto-detect it.

**Comment by nfeltrin7s** [comment]
My use case is within:
- you use Terragrunt where plugin caches are decoupled from their directories - see hashicorp/terraform-ls#32 (comment)

I tried disabling the language server as a workaround, but it still produces the error message at startup. Is that expected? Is there any other way to disable the LS other than by using the command in the interface (configuration files etc)?

**Comment by paultyng** [comment]
FYI, the downgrading issue is an upstream VS Code issue (see https://github.com/microsoft/vscode/issues/99699#issuecomment-643379024). We will track the upstream in #385 and provide updates there.

**Comment by mieseprem** [comment]
Hello folks,

yesterday I ran into this issue and found the reason in the way I work.
In VSCode my primary shell is WSL Shell. Outside of VS Code I always use Ubuntu WSL Shell.
If I initialize my terraform project plugins are stored in .terraform/plugins/linux_amd64/
On VS Code startup terraform-ls is executed as windows executable. It doesn't take the linux_amd64 directory into account. 
My workaround was to initialize the project a seccond time out of a windows console with a terraform windows binary. After that a second folder unter plugins/ was created (.terraform/plugins/windows_amd64). Now startup of terraform-ls in VS Code don't crash anymore.

**Comment by nhart-alf** [comment]
> 
> 
> Hello folks,
> 
> yesterday I ran into this issue and found the reason in the way I work.
> In VSCode my primary shell is WSL Shell. Outside of VS Code I always use Ubuntu WSL Shell.
> If I initialize my terraform project plugins are stored in .terraform/plugins/linux_amd64/
> On VS Code startup terraform-ls is executed as windows executable. It doesn't take the linux_amd64 directory into account.
> My workaround was to initialize the project a seccond time out of a windows console with a terraform windows binary. After that a second folder unter plugins/ was created (.terraform/plugins/windows_amd64). Now startup of terraform-ls in VS Code don't crash anymore.

This worked for me also, running WSL as my default terminal within VS Code.

**Comment by J0F3** [comment]
@mieseprem Btw. Also the [Remote - WSL ](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl) extension may help you a lot with this workflow. It allows basically to run the whole VS Code in the context of WSL and so you would not need to run a second init in Windows nor you need the Windows Terraform executable at all.
I can really recommend it, especially together with the Terraforming extension.

**Comment by tb102122** [comment]
> I found a simple workaround that works well for me. I added a single `terraform-ls.tf` file at the root of the project with only a list of providers I use.
> 
> UPDATE: you need to run `terraform init` from the root of your project (the server won't do it for you)

quick and simple workaround! Thanks for the hint.

**Comment by WodansSon** [comment]
I too hit this issue, the way I worked around it was to copy a random configuration from the (examples)[https://github.com/terraform-providers/terraform-provider-azurerm/tree/b-frontdoor-timeout/examples/api-management] folder into the root of my github repro forked directory, opened a windows command prompt to that folder and ran the `terraform init` command. Once it was initialized I simply deleted the files I copied over. Once I opened VSCode again the issue was fixed. Looks like it just needed to bin place the new runtime for VSCode to pick up the new bits or something along those lines. Would be nice if the VSCode extension did this bit for you, but a pretty simple workaround. HTH 🙂

**Comment by moosahmed** [comment]
The workaround I used and it works.

I copied a `.terraform` dir that has the `plugins/darwin_amd64/lock.json` from a subdirectory that had it and pasted it in the root directory I am opening in vscode.

This solved the issue. But is a horrible workaround.

**Comment by mislavcimpersak** [comment]
Reporting another horrible workaround:

made a symlink in the root dir of my project:

`ln -s ./infrastructure/.terraform .terraform`

My dir structure:
```
.
├── .git
├── .terraform -> ./infrastructure/.terraform
├── .vscode
│   └── settings.json
├── infrastructure
|   ├── .terraform
|   ├── main.tf
└── normal project stuff
```

**Comment by aaronsteers** [comment]
Hi, Terraform. Can we ***please*** put a note in the Extension readme that explicitly states that multi-folder workspaces are not supported - along with any other caveats of expected and not supported folder structures? Would have saved me 2 hours today, probably hundreds (thousands?) of hours per month across all users.

Thank you!!

**Comment by aaronsteers** [comment]
Fwiw, I still haven't gotten the features to work properly but it sounds like that's because in addition to using multi-folder workspaces exclusively, I'm also using non-traditional folder structures. 🤦‍♂️

I'm _strategically_ deciding to just give up for now and live without the autocomplete and syntax-highlighting goodness, but will watch this thread for resolution.

**Comment by aaronsteers** [comment]
> Hi, Terraform. Can we _**please**_ put a note in the Extension readme that explicitly states that multi-folder workspaces are not supported - along with any other caveats of expected and not supported folder structures? Would have saved me 2 hours today, probably hundreds (thousands?) of hours per month across all users.
> 
> Thank you!!

Proposal for review: https://github.com/hashicorp/vscode-terraform/pull/430

**Comment by taylorsmcclure** [comment]
I really hope this gets merged. I have a workspace with a ton of separate terraform repos in various different paths on my local machine, and I have to do the "shuffle the directory to the top" dance to get it to work properly.

**Comment by radeksimko** [comment]
Hi folks,
the issue is not addressed _yet_, but we have a plan for addressing it and I'm just converging all the issues with the same root cause, such as this one, into a single issue: https://github.com/hashicorp/vscode-terraform/issues/569

Thank you for understanding and patience.
